shards/sched.go

```diff
@@ -0,0 +1,542 @@
+// process acquires the full semaphore. This is equivalent to how concurrency
+// is managed in upstream. It exists as a fallback while we test
+// multiScheduler.
+type semaphoreScheduler struct {
+	throttle *semaphore.Weighted
+	capacity int64
+}
+
+// Acquire implements scheduler.Acquire.
+func (s *semaphoreScheduler) Acquire(ctx context.Context) (*process, error) {
+	return s.acquire(ctx, 1)
+}
+
+// Exclusive implements scheduler.Exclusive.
+func (s *semaphoreScheduler) Exclusive() *process {
+	// Won't error since context.Background won't expire.
+	proc, _ := s.acquire(context.Background(), s.capacity)
+	return proc
+}
+
+func (s *semaphoreScheduler) acquire(ctx context.Context, weight int64) (*process, error) {
+	if err := s.throttle.Acquire(ctx, weight); err != nil {
+		return nil, err
+	}
+	return &process{
+		releaseFunc: func() {
+			s.throttle.Release(weight)
+		},
+	}, nil
+}
+
+// process represents a running search query or an exclusive process. When the
+// process is done a call to Release is required.
+type process struct {
+	// yieldTimer ensures we only call yieldFunc once after a deadline.
+	yieldTimer *deadlineTimer
+	// yieldFunc is called once by Yield.
+	yieldFunc func(context.Context) error
+
+	// releaseFunc is called once by Release
+	releaseFunc func()
+}
+
+// Release the resources/locks/semaphores associated with this process. Can
+// only be called once.
+func (p *process) Release() {
+	if p.yieldTimer != nil {
+		p.yieldTimer.Stop()
+	}
+
+	p.releaseFunc()
+}
+
+// Yield may block to allow another process to run. This should be called
+// relatively often by a search to allow other processes to run. This can not
+// be called concurrently.
+//
+// The only error it will return is a context error if ctx expires. In that
+// case the process should stop running and call Release.
+func (p *process) Yield(ctx context.Context) error {
+	if p.yieldTimer == nil || !p.yieldTimer.Exceeded() {
+		return nil
+	}
+
+	// Try to yield. This can return an error if our context expired.
+	err := p.yieldFunc(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Successfully yielded. Stop our timer and mark it nil so we don't call
+	// yieldFunc again.
+	p.yieldTimer.Stop()
+	p.yieldTimer = nil
+
+	return nil
+}
+
+// newDeadlineTimer returns a timer which fires after deadline. Once it fires
+// Exceeded will always return true. Callers must call Stop when done to
+// release resources.
+func newDeadlineTimer(deadline time.Time) *deadlineTimer {
+	return &deadlineTimer{
+		t: time.NewTimer(deadline.Sub(time.Now())),
+	}
+}
+
+type deadlineTimer struct {
+	// t.C fires after deadline. Once it fires we set to nil to indicate it has
+	// fired.
+	t *time.Timer
+}
+
+// Exceeded returns true if time is after the deadline.
+func (t *deadlineTimer) Exceeded() bool {
+	if t.t == nil {
+		return true
+	}
+	select {
+	case <-t.t.C:
+	default:
+		return false
+	}
+
+	t.Stop()
```

Comment:
Why do we call `Stop`?

Reply:
`t.t.C` won't fire again. By calling Stop we end up setting `t.t` to nil so in future calls to exceeded we return true.